### ISSUE TITLE
Fix unit tests (was: Broken unit tests on OS X)

### ISSUE THREAD

**Issue by jaeandersson** [jaeandersson/swig#6]

**Comment by jaeandersson** [comment]
@ibell @KrisThielemans @tesch1 I finally got the unit test framework working but I get loads and loads of errors. The list above isn't even complete. How do you usually run the unit tests, via "make check-matlab-test-suite" from the command line? Do all the unit tests not marked as broken in Examples/test-suite/matlab/Makefile.in pass for you? It might be a OS X thing (I'm on OS X 10.10 with Clang++ and MATLAB 2014B).

**Comment by tesch1** [comment]
(with the changes https://github.com/tesch1/swig/commit/86695f1a41eeb5169684ed373e7e5b49e56cc855)
MacOS 10.9.5

```
$ make  check-matlab-test-suite RUNPIPE=\>\>/tmp/xxx
$ grep skipping /tmp/xxx | wc -l
     306
$ grep failed /tmp/xxx | wc -l
      53
$ grep passed /tmp/xxx | wc -l
      63
$ matlab | grep R201
                             R2011b (7.13.0.564) 64-bit (maci64)
$ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 6.0 (clang-600.0.54) (based on LLVM 3.5svn)
Target: x86_64-apple-darwin13.4.0
Thread model: posix
```

306 that aren't even tried!!!

Here's the difference against your list... I can send you the log with the detailed errors if you want to look at it, there's no nice way to attach it here, and it seems kind of spammy to paste the whole 80k file into a comment.

```
$ diff -y -W 70 mtfailed jafailed 
array_member                      |     abstract_typedef2
                                  >     abstract_virtual
                                  >     allprotected_not
                                  >     arrays_global
                                  >     char_strings
                                  >     class_ignore
                                  >     constructor_copy
contract                                contract
cpp_enum                                cpp_enum
cpp_namespace                           cpp_namespace
default_constructor                     default_constructor
                                  >     dynamic_cast
exception_order                         exception_order
friends                                 friends
                                  >     fvirtual
grouping                                grouping
import_nomodule                         import_nomodule
imports                           |     inherit_missing
li_attribute                            li_attribute
li_carrays                              li_carrays
li_carrays                        |     member_funcptr_galore
li_cmalloc                        |     member_pointer
li_cpointer                       <
li_cpointer                       <
li_std_vector                     <
minherit                                minherit
mod                               |     namespace_spaces
multi_import                      <
namespace_typemap                       namespace_typemap
nested_structs                    |     namespace_virtual_method
null_pointer                            null_pointer
overload_extend                         overload_extend
overload_extend                   <
overload_rename                         overload_rename
overload_simple                         overload_simple
overload_template                       overload_template
preproc                           <
preproc_constants                       preproc_constants
                                  >     refcount
reference_global_vars                   reference_global_vars
return_const_value                <
smart_pointer_extend              <
smart_pointer_member              <
smart_pointer_multi               <
smart_pointer_multi_typedef       <
smart_pointer_not                 <
smart_pointer_overload            <
smart_pointer_rename              <
smart_pointer_simple              <
smart_pointer_templatevariable    <
smart_pointer_typedef             <
struct_value                      <
template_default_arg                    template_default_arg
template_static                   |     template_inherit
template_type_namespace           |     template_specialization_defarg
template_typedef                  <
template_typedef_cplx             <
template_typedef_cplx2                  template_typedef_cplx2
template_typedef_import           |     typedef_inherit
unions                            |     using_private
virtual_poly                            virtual_poly
voidtest                                voidtest
```

**Comment by KrisThielemans** [comment]
indeed. a lot of the test break at compilation stage, also on Linux. I believe I indicated accordingly on swig-devel.Quite a lot of them break because of directors, which are currently not supported. William has indicated however that he wants the code to compile cleanly anyway (I guess returning errors to matlab then). There's also a few others which get repeatedly flagged-up. So, I'm somewhat optimistic that with a small amount of work, the fraction of tests that will work at compilation stage will increase dramatically.

The "skipping" means that the test is only compiled, but there is no corresponding runme to do additional tests in matlab itself.

A large fraction of the runme.m come straight from conversion from the Octave ones with an automatic process. As the commit-log https://github.com/jaeandersson/swig/commit/b74e35eb9b9c28fb811ae865d3147c1bed2b26bc says some of these still need manual editing, e.g. because they use Octave's in-file function support.

There's also a large fraction of the tests that don't have a runme, but I guess not more than octave.

Finally, I had problems with R2014a on Linux actually running the tests themselves. When called from make, matlab crashed (but when executing the test manually, it all worked fine). That seemed to be a matlab issue, so I hadn't reported this one on the list yet and I haven't had time to investigate this further.

so... happy bug fixing!

**Comment by jaeandersson** [comment]
@tesch1 @KrisThielemans Thanks a lot, that makes it clearer. It appears as if many (not all) of the tests fail more violently for me, with segfaults in MATLAB from which the test suite cannot recover. I will try to get MATLAB up and running in my Ubuntu Virtualbox where I have better debuggers. So yeah, bugfixing...

**Comment by tesch1** [comment]
Strangely, I had the same problem on Linux -- tests that crash and hang the test-suite because MATLAB throws up a prompt at the crash.  My linux matlab version is newer though, so it might have more to do with particular matlab versions' handling errors.

**Comment by tesch1** [comment]
here's some - https://github.com/tesch1/swig/commit/b219d4ee46f35c170496fa34118bae94585e87cc

**Comment by ibell** [comment]
I had to disable matlab testing for my library because on windows on errors
matlab would go interactive, and if you are always throwing errors, you end
up with one copy of the interpreter per test, which very quickly eats up
whatever RAM you can throw at it.

**Comment by KrisThielemans** [comment]
yes, I don't know how to force matlab to exit on certain errors.

How do you run these on Windows by the way?

**Comment by ibell** [comment]
Nope, not yet, what's the protocol for running tests again?  I can set up
my windows box to run the tests.  Do you use buildbot?

**Comment by jaeandersson** [comment]
Note that eventually, we need to get the unit tests running on travis-ci, since that's what SWIG uses. In our project, we're also migrating from buildbot to travis-ci, though for details you have to ask @jgillis.

**Comment by tesch1** [comment]
Will it be enough to have the tests just build?  I dont see any way to run MATLAB under travis-ci...(?)

**Comment by KrisThielemans** [comment]
Doesn’t travis just use the existing make+configure stuff?

This is a conversation for swig-devel really.

**Comment by jaeandersson** [comment]
Current situation for me (Virtualbox with Ubuntu 14.04 64 bit guest, gcc-4.7, MATLAB 2014b):
- passed: 25 tests
- failed: 14 tests
- skipped: 123 tests

Only tests until `overload_simple` considered, at that point MATLAB segfaults.

**Comment by jaeandersson** [comment]
Test framework running much more smoothly on my Mac and Clang, but the tests are only compiled, not run. Probably an issue with my setup - I have a new OS installation. Anyway, I don't get segfaults all the time anymore and the compilation errors I get look mostly harmless, as in missing features (in particular directors) or MATLAB symbols commencing with underscore.
Will try again in Ubuntu and GCC.

**Comment by KrisThielemans** [comment]
not sure where those underscores can come from. I don't see that at all on Ubuntu.

If the tests aren't running matlab, I guess your matlab wasn't found properly at configure time?

**Comment by jaeandersson** [comment]
Not sure, could be the MATLAB version perhaps (I use 2015a). Anyway, I'll try to get at least the generation step working in the next couple of weeks - should be pretty straightforward.

**Comment by jaeandersson** [comment]
I get warnings like this:

```
Warning: invalid MATLAB symbol '__str__' (function)
  Symbols may not start with '_'.  Maybe try something like this: %rename(u__str__) __str__;
Warning: invalid MATLAB symbol '__str__' (function)
  Symbols may not start with '_'.  Maybe try something like this: %rename(u__str__) __str__;
```

But it looks like a legitimate warning.

**Comment by jaeandersson** [comment]
I get the following error as well:

```
partialchecking matlab testcase template_default2
*** Internal error. Inconsistent clientdata for type 'oss::Module< oss::traits,oss::UnaryPolarization,oss::Interface_tpl< oss::UnaryPolarization > > *'
*** '"template_default2.Module_UP2"' != '"template_default2.Module_UP1"'
swig: Swig/typesys.c:1629: SwigType_remember_clientdata: Assertion `0' failed.
make[4]: *** [matlab_cpp] Aborted (core dumped)
make[3]: *** [template_default2.cpptest] Error 2
partialchecking matlab testcase template_default_arg
```

@wsfulton - any idea what this could be about?

**Comment by jaeandersson** [comment]
And this:

```
partialchecking matlab testcase template_specialization_defarg
*** Internal error. Inconsistent clientdata for type 'C< double,double > *'
*** '"template_specialization_defarg.C_d"' != '"template_specialization_defarg.C_dd"'
swig: Swig/typesys.c:1629: SwigType_remember_clientdata: Assertion `0' failed.
make[4]: *** [matlab_cpp] Aborted (core dumped)
make[3]: *** [template_specialization_defarg.cpptest] Error 2
```

**Comment by KrisThielemans** [comment]
the **str** error is because some of the tests define a **str** function (it is used for instance in python where this function is used to provide a pretty-print mechanism). MATLAB doesn't allow such names and so the run-time test fails. This might be a topic for the swig-devel mailing list on how to handle this best.

the clientdata errors are tougher. They have to do with how we store the type info in the clientdata. I think this has to do with the search for base classes, but I'm not sure (and don't have the time to help right now)

**Comment by jaeandersson** [comment]
OK, sent an email to swig-devel.

**Comment by wsfulton** [comment]
I haven't reviewed any changes in this branch, but sounds like the core runtime might have changed or the symbol tables are modified in some way (neither of which is advisable unless you know what you are doing or run the changes over all the other languages first).

**Comment by jaeandersson** [comment]
I am not aware of any changes to the core runtime. I guess that should be detected if I do a `diff` with master? I could try to pull the latest changes to master and see if the problem remains. Maybe we simply pulled a bad master revision.

**Comment by KrisThielemans** [comment]
I don't think the SwigType_remember_clientdata problems have anything to do with master, but are a matlab problem how we handle the clientdata.

After fixing the usage of error() (see https://github.com/jaeandersson/swig/issues/30), my current status on Linux, MATLAB 2012a, gcc 4.6.3:

> $ make check-matlab-test-suite >& check.log
> $ grep skipped check.log | wc -l
> 0
> $ grep passed check.log | wc -l
> 59
> $ grep failed check.log | wc -l
> 121

still some director errors of course, the usage of _ in the test-case mentioned above, non-existent SWIG_Matlab_ConvertPacked and SWIG_Matlab_NewPackedObj. After that, it starts to become harder...

**Comment by KrisThielemans** [comment]
With Pull Request #41, the commit on my fork mentioned in Issue #30  and the fix in Issue #39, the status becomes a bit better:

> $ grep passed check.log | wc -l
> 80
> $ grep failed check.log | wc -l
> 58

Large remaining ones are SWIG_Matlab_ConvertPacked, usage of swig_this and some generated code that uses **deref** etc

**Comment by jaeandersson** [comment]
Tests still do not run until completion for me (last one tested is `multi_import`. But I also get 80 passed tests, along with 47 failed ones.

**Comment by KrisThielemans** [comment]
I thought multi_import was the last one. It's the last of MULTI_CPP_TEST_CASES, which is the last of NOT_BROKEN_TEST_CASES as we don't have EXTRA_TEST_CASES. no?

**Comment by jaeandersson** [comment]
I see, that's good.

**Comment by KrisThielemans** [comment]
a closer look by running e.g. partialcheck shows that when a runme.m doesn't exist, then the Makefile doesn't write "failed" or "passed" (see run_testcase in test-suite/matlab/Makefile). So we're having quite a lot more compile-only tests that work. This seems somewhat confusing though. Not sure how other languages handle this. Python seems to report if there's a run_test or not.

also a bit strange is that we have 160 runme.m files but less tests that report something. probably there's some files that I copied from Octave that aren't in common.mk.

**Comment by jaeandersson** [comment]
If I run partialcheck I only get one test failing in an ugly way:

```
MacBook-Pro:matlab jaeandersson$ make -s template_specialization_defarg.cpptest
checking matlab testcase template_specialization_defarg
*** Internal error. Inconsistent clientdata for type 'C< double,double > *'
*** '"template_specialization_defarg.C_d"' != '"template_specialization_defarg.C_dd"'
Assertion failed: (0), function SwigType_remember_clientdata, file ../../Source/Swig/typesys.c, line 1629.
make[1]: *** [matlab_cpp] Abort trap: 6
make: *** [template_specialization_defarg.cpptest] Error 2
```

@wsfulton - any idea what could be the reason for this?

**Comment by wsfulton** [comment]
That looks like it is in the runtime system. I'm afraid you'll have to debug it. Make sure you have the latest from master and you have no changes in your branch compared to master outside of matlab.cxx. A new module may also have a couple of changes in swigmain.cxx.

**Comment by KrisThielemans** [comment]
As I mentioned before, I checked and we had no changes in SWIG code last time I checked. This error has to do with how we store type type-name-string in the clientdata (matlab.cxx L1986). 
If I look at template_specialization_defarg.i, I see

> //
> // This one fails
> //
> %template(C_dd) C<double,double>;
> %template(C_d) C<double>;

(Note the interesting comment. Should that be there?) 
I infer from this that the problem is that the same class gets wrapped twice (as C has double as default 2nd template arg). Indeed, I can reproduce this with the following interface file

> %module doublename
> %inline %{
>   template <class T> class A {
>   public:
>     int a;
>   };
>   %}
> %template(A_d) A<double>;
> %template(A2_d) A<double>;

Which gives me the corresponding error

>  **\* Internal error. Inconsistent clientdata for type 'A< double > _'
> *_\* '"doublename.A2_d"' != '"doublename.A_d"'

Of course, it seems strange to wrap the same class with 2 different names. @wsfulton, what is the SWIG policy for this? If it is allowed, how would we handle this? I suppose we could let the Matlab module throw an error if it detects this happening (but how?), and put in a conditional in the above test-case.
By the way, maybe we should add an explicit test-case for this?

**Comment by KrisThielemans** [comment]
@jaeandersson  I've pushed a commit to my fork to change the echo statements when running the test-suite (check the log message). I'm not so happy with the formatting, but at least now we can count the "compile-only" tests as well

> $ fgrep passed check.log|fgrep passed |fgrep runme|wc -l
> 339
> $ fgrep passed check.log|fgrep passed |fgrep -v runme|wc -l
> 80

which sounds better than before :-)

I'm afraid I didn't bother generating a pull-request for it. sorry

**Comment by KrisThielemans** [comment]
@jaeandersson , I've pushed another commit to my fork fixing some runme files. See https://github.com/KrisThielemans/swig/commit/9ce53c42e043d11c6ca2ec6049fc8b5ee37a08a4 
some of these fixes now reveal actual problems as opposed to just Octave copy-paste, so still only 83 tests passed.

**Comment by jaeandersson** [comment]
I've managed to isolate the remaining partialcheck failure to the code added in 08718058209639eff8c3636ce4b09be521699d62.

**Comment by jaeandersson** [comment]
From what I understand the partialcheck failure is related NewPointerObj. So I guess the proper thing to do is to do a thorough refactoring of SWIG_Matlab_NewPointerObj. The current design isn't clean.

**Comment by KrisThielemans** [comment]
I get a MATLAB segmentation violation with the dynamic_cast test-case after this.

**Comment by jaeandersson** [comment]
@KrisThielemans  Can you double-check that the offending commit is 3a184936f3af565dd78f256c9fc3c337f6289124 and not some earlier commit?

**Comment by KrisThielemans** [comment]
This will be completely outdated after #96 . I recommend closing it.
